ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -548,5 +552,41 @@ private void addDatasetVersion(BlobProperties blobProperties, RestRequest restRe
             RestServiceErrorCode.getRestServiceErrorCode(ex.getErrorCode()));
       }
     }
+
+    /**
+     * Support delete the dataset version out of retentionCount
+     */
+    private void deleteDatasetVersionOutOfRetentionCount() {
+      Dataset dataset = (Dataset) restRequest.getArgs().get(RestUtils.InternalKeys.TARGET_DATASET);
+      try {
+        if (RestUtils.isDatasetVersionQueryEnabled(restRequest.getArgs())) {
+          String accountName = dataset.getAccountName();
+          String containerName = dataset.getContainerName();
+          String datasetName = dataset.getDatasetName();
+          List<DatasetVersionRecord> datasetVersionRecordList =
+              accountService.getAllValidVersionsOutOfRetentionCount(accountName, containerName, datasetName);
+          for (DatasetVersionRecord record : datasetVersionRecordList) {
+            String version = record.getVersion();
+            RequestPath requestPath = getRequestPath(restRequest);
+            RequestPath newRequestPath = new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(),
+                requestPath.getPathAfterPrefixes(),
+                NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH + version,
+                requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
+            // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
+            restRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+            restRequest.setArg(InternalKeys.TARGET_ACCOUNT_KEY, null);
+            restRequest.setArg(InternalKeys.TARGET_CONTAINER_KEY, null);
+            try {
+              deleteBlobHandler.handle(restRequest, restResponseChannel, null);
+            } catch (Exception e) {
+              //do not throw the exception or handle it since we don't care if it has been deleted successfully.
+              LOGGER.error("Failed to delete the datasetVersionRecord :" + record);
+            }
+          }
```

Comment:
oh, this is probably not going to work.

You see, the delete handler is an async method, so it will not finished the execution before handle method returns. And if we have multiple versions to delete, then we would override the RequestPath in this for loop since the they share the same restRequest.

For example:
Say we have version1 and version2 to delete.
for version1, we replace RequestPath in restRequest to version1 and call deleteBlobHandler.handle(restRequest).
Before it finishes, we move on to version2, and now we replace the RequestPath in restRequest and call deleteBlobHandler.handle(restRequest) again.
But since deleting version 1 isn't done yet, any logic in deleting version1 would now see version2.

We might have to create a new restRequest for each version.

Reply:
Yeah, I feel it's hard to clone a restRequest, so I decided to use the recursive callback to delete dataset version synchronously, and ignore the delete failures. Normally there would be only one version need to be deleted.